File: php/src/Client/Shell/InputReader.php
<?php

namespace Shapes\Client\Shell;

function readInput(): string {
    echo "Your Answer >>> ";
    $user_input = fgets(STDIN);
    $user_input = FALSE !== $user_input ? strtolower(trim($user_input)) : "";
    readExit($user_input);

    return $user_input;
}

function readNumber(): float {
    while (TRUE) {
        $user_input = readInput();

        if (is_numeric($user_input)) {
            return (float) $user_input;
        }
        echo "Invalid number. Please try again." . PHP_EOL;
    }
}

function readExit(string $user_input): void {
    if ("exit" === $user_input || "e" === $user_input) {
        echo "Thank you for using the Shapes CLI!" . PHP_EOL;

        exit(0);
    }
}

File: cplusplus/src/client/shell/input-reader.cpp
#include <boost/algorithm/string.hpp>
#include <iostream>
#include <print>
#include <shapes/client/shell.hpp>

using namespace boost::algorithm;
using namespace std;

namespace shapes::client::shell {
namespace {
    inline void readExit(const string& userInput) {
        if (userInput == "exit" || userInput == "e") {
            std::println("Thank you for using the Shapes CLI!");
            std::quick_exit(0);
        }
    }
} // namespace

auto readInput() -> string {
    std::print("Your Answer >>> ");
    string userInput;
    getline(cin, userInput);
    userInput = to_lower_copy(trim_copy(userInput));
    readExit(userInput);
    return userInput;
}

auto readNumber() -> double {
    while (true) {
        const string userInput = readInput();

        try {
            return stod(userInput);
        } catch (const invalid_argument&) {
            std::println("Invalid number. Please try again.");
        }
    }
}
} // namespace shapes::client::shell
